vuepress/docs/_posts/2023-03-17-building-business-based-autonomous-components.md

```diff
@@ -0,0 +1,192 @@
+
+
+# Architectures around Domain Model
+
+There are many Domain Centric Software Architectures. If they all focus on the Domain Model protection, depending on your needs, one may be more suited than the other.
+
+I’m mainly a Java developer. On the backend, my “default” Domain Centric architecture is a peculiar flavor of [Ports and Adapters](https://alistair.cockburn.us/hexagonal-architecture/) (aka Hexagonal Architecture).
+
+In this flavor, of course I’ll start with the Domain Model at the center: 
+
+
+![drawing](./images/2023-03-17-image1.png)
+
+Most of the time, my goal will be to expose Domain operations through WebServices. This is the Primary Adapters job:
+
+
+
+![drawing](./images/2023-03-17-image2.png)
+
+This is where the flavor specificity kicks in: The Application Layer. This is a very thin layer mainly composed by XXXApplicationService that are:
+
+
+
+* Wiring transactions management;
+* Wiring authorization control;
+* Doing really simple orchestration.
+
+This Layer has Framework dependencies so it will be really easy to add the wiring capabilities.
+
+Since the Application Layer needs to interact with infrastructure we’ll need to add ports & adapters:
+
+
+
+![drawing](./images/2023-03-17-image3.png)
+
+The ports are interfaces in the domain and the secondary adapters are implementations of these ports.
+
+> Ports are in the Domain because it is really common to inject them in the domain, ex: in Domain Services.
+
+I really like this architecture because each part is only doing one thing so we can focus on doing it well:
+
+* Primary adapters only need to make a great adaptation. For example, they can have great REST documentation without polluting anything else;
+
+* Application is really thin, no real Business value here;
+
+* Domain is the code that really matters so we can focus on it’s correctness and expressivity;
+
+* Secondary adapters only need to make a great adaptation to infrastructure elements. For example, we can focus on having a really performant data model for our business cases here.
+
+> This very quick introduction is probably not enough to be able to code in this kind of architecture. You can have a look at [this live coding](https://www.youtube.com/watch?v=mEECPRZjajI) to see it in action.
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/mEECPRZjajI" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>
+
+If you are using OOP you should give it a try. If this doesn’t suit your needs you might want to have a look at [Onion Architecture](https://jeffreypalermo.com/2008/07/the-onion-architecture-part-1/) or [Functional Core Imperative Shell](https://kennethlange.com/functional-core-imperative-shell/) if your Domain Model is in FP.
```

Comment:
```suggestion
If you are using OOP you should give (be explicit about what) a try. If this doesn’t suit your needs you might want to have a look at [Onion Architecture](https://jeffreypalermo.com/2008/07/the-onion-architecture-part-1/) or [Functional Core Imperative Shell](https://kennethlange.com/functional-core-imperative-shell/) if your domain model is in FP.
```

Reply:
@jyunmitch the "it" here refers to the hexagonal architecture in the sense that, if you are using OOP, try to implement the hexagonal architecture because it worth it.